File: packages/sdk-go/sharded_limiter.go
package rateguard

import (
	"context"
	"math"
	"sync"
	"sync/atomic"
	"time"
)

// shardCount partitions the key space so bucket lookups on different keys
// never contend on the same lock. Must be a power of two.
const shardCount = 64

// ShardedLimiter is a lock-free, sharded token bucket limiter.
//
// It makes the same decisions as MemoryLimiter (identical token bucket math,
// identical Retry-After rounding) but removes both contention points of the
// mutex design:
//
//   - the global cache mutex becomes 64 shard mutexes, held only for the
//     map lookup — different keys almost never contend;
//   - the per-bucket mutex disappears entirely: bucket state is a single
//     atomic int64 updated with a CAS loop, so admissions on one hot key
//     proceed without any lock (the pattern proven by uber-go/ratelimit's
//     atomicInt64Limiter).
//
// The whole bucket state — (tokens, last refill time) — is encoded losslessly
// as one number: fullAtNanos, the instant the bucket would refill completely.
//
//	tokens(now) = burst − max(0, fullAt − now) × rps / 1e9
//	consume 1   = fullAt' = now + (burst − (tokens−1)) / rps × 1e9
//
// The zero value (fullAt = 0, i.e. the distant past) encodes a full bucket,
// so freshly created buckets need no initialization write. An idle bucket
// refills to full by construction, which subsumes MemoryLimiter's 10-minute
// idle reset.
type ShardedLimiter struct {
	shards [shardCount]limiterShard
	clock  Clock
}

type limiterShard struct {
	mu      sync.Mutex
	buckets *boundedCache[string, *atomicBucket]
}

// atomicBucket holds fullAtNanos. Padding keeps a hot bucket on its own cache
// line so CAS traffic does not false-share with neighboring allocations
// (same layout rationale as uber-go/ratelimit).
type atomicBucket struct {
	_      [64]byte //nolint:unused // cache-line padding
	fullAt atomic.Int64
	_      [56]byte //nolint:unused // cache-line padding
}

// NewShardedLimiter creates the lock-free in-process limiter used by default.
func NewShardedLimiter() *ShardedLimiter {
	return newShardedLimiterWithClock(systemClock{}, defaultMemoryLimiterCacheCapacity)
}

func newShardedLimiterWithClock(clock Clock, capacity int) *ShardedLimiter {
	if clock == nil {
		clock = systemClock{}
	}
	if capacity <= 0 {
		capacity = defaultMemoryLimiterCacheCapacity
	}

	perShard := capacity / shardCount
	if perShard <= 0 {
		perShard = 1
	}

	l := &ShardedLimiter{clock: clock}
	for i := range l.shards {
		l.shards[i].buckets = newBoundedCache[string, *atomicBucket](perShard)
	}
	return l
}

// shardFor hashes the key with FNV-1a (inlined: zero allocations) and masks
// into the shard array.
func (l *ShardedLimiter) shardFor(key string) *limiterShard {
	const (
		offset64 = 14695981039346656037
		prime64  = 1099511628211
	)
	var h uint64 = offset64
	for i := 0; i < len(key); i++ {
		h ^= uint64(key[i])
		h *= prime64
	}
	return &l.shards[h&(shardCount-1)]
}

func (l *ShardedLimiter) bucketFor(key string, create bool) *atomicBucket {
	shard := l.shardFor(key)
	shard.mu.Lock()
	defer shard.mu.Unlock()

	if !create {
		bucket, _ := shard.buckets.get(key)
		return bucket
	}
	return shard.buckets.getOrCreate(key, func() *atomicBucket {
		return &atomicBucket{} // zero value = full bucket
	})
}

// tokensAt reports the current token count encoded by fullAt at time now.
func tokensAt(fullAtNanos, nowNanos int64, policy PolicyPreset) float64 {
	deficit := fullAtNanos - nowNanos
	if deficit <= 0 {
		return float64(policy.Burst)
	}
	tokens := float64(policy.Burst) - float64(deficit)/1e9*float64(policy.RequestsPerSecond)
	if tokens < 0 {
		return 0
	}
	return tokens
}

// denyDecision mirrors MemoryLimiter's Retry-After rounding exactly:
// whole seconds, ceil((1 − tokens) / rps).
func denyDecision(tokens float64, policy PolicyPreset) AdmissionDecision {
	retry := time.Duration(math.Ceil((1.0-tokens)/float64(policy.RequestsPerSecond)) * float64(time.Second))
	if retry < 0 {
		retry = time.Second
	}
	return AdmissionDecision{
		Allowed:    false,
		Applied:    true,
		Remaining:  0,
		RetryAfter: retry,
		Limit:      policy.RequestsPerSecond,
	}
}

func (l *ShardedLimiter) Allow(_ context.Context, key string, policy PolicyPreset) (AdmissionDecision, error) {
	if policy.RequestsPerSecond <= 0 || policy.Burst <= 0 {
		return AdmissionDecision{Allowed: true, Applied: false, Remaining: -1, Limit: -1}, nil
	}

	clock := l.clock
	if clock == nil {
		clock = systemClock{}
	}
	now := clock.Now().UTC().UnixNano()

	bucket := l.bucketFor(key, true)

	for {
		fullAt := bucket.fullAt.Load()
		tokens := tokensAt(fullAt, now, policy)

		if tokens < 1 {
			// Denials do not consume and need no state write: the deficit
			// encoded in fullAt already represents the refilled state.
			return denyDecision(tokens, policy), nil
		}

		newTokens := tokens - 1
		newFullAt := now + int64((float64(policy.Burst)-newTokens)/float64(policy.RequestsPerSecond)*1e9)

		if bucket.fullAt.CompareAndSwap(fullAt, newFullAt) {
			remaining := int(math.Floor(newTokens))
			if remaining < 0 {
				remaining = 0
			}
			return AdmissionDecision{
				Allowed:   true,
				Applied:   true,
				Remaining: remaining,
				Limit:     policy.RequestsPerSecond,
			}, nil
		}
		// Lost the race to a concurrent admission — reload and retry.
	}
}

// Peek reports what Allow would decide right now without consuming a token.
// It never creates bucket state for unseen keys.
func (l *ShardedLimiter) Peek(_ context.Context, key string, policy PolicyPreset) (AdmissionDecision, error) {
	if policy.RequestsPerSecond <= 0 || policy.Burst <= 0 {
		return AdmissionDecision{Allowed: true, Applied: false, Remaining: -1, Limit: -1}, nil
	}

	clock := l.clock
	if clock == nil {
		clock = systemClock{}
	}
	now := clock.Now().UTC().UnixNano()

	bucket := l.bucketFor(key, false)
	if bucket == nil {
		return AdmissionDecision{
			Allowed:   true,
			Applied:   true,
			Remaining: policy.Burst,
			Limit:     policy.RequestsPerSecond,
		}, nil
	}

	tokens := tokensAt(bucket.fullAt.Load(), now, policy)
	if tokens < 1 {
		return denyDecision(tokens, policy), nil
	}

	return AdmissionDecision{
		Allowed:   true,
		Applied:   true,
		Remaining: int(math.Floor(tokens)),
		Limit:     policy.RequestsPerSecond,
	}, nil
}

File: packages/sdk-node/src/core/sharded-limiter.ts
/**
 * Sharded token-bucket limiter — Node port of Go's ShardedLimiter
 * (packages/sdk-go/sharded_limiter.go).
 *
 * HONEST SCOPE — read before reaching for this over `RateLimiter`:
 *
 * Go's ShardedLimiter exists to remove lock contention: 64 shard mutexes
 * (instead of one global mutex) plus a per-bucket atomic CAS loop so hot-key
 * admissions never block on a lock at all — a real win under concurrent
 * OS-thread access (the pattern uber-go/ratelimit proved).
 *
 * A single Node.js process runs JS on ONE thread. There is no OS-thread
 * contention for a lock-free structure to remove here — two "concurrent"
 * calls to `increment()` never actually race, because the event loop runs
 * each synchronous call to completion before the next one starts. Porting
 * the CAS-retry-loop shape to Node would just be decoration around a plain
 * read-modify-write; this file does not pretend otherwise.
 *
 * So why does this class exist in Node at all? Two honest reasons:
 *
 *   1. Decision parity. This is the SAME token-bucket math as `RateLimiter`,
 *      encoded the way Go encodes it (a single "bucket would be full at this
 *      instant" timestamp, `fullAt`, instead of a {tokens, lastRefill} pair).
 *      Cross-language conformance vectors (conformance/token_bucket_vectors.json)
 *      can replay against this implementation and assert it produces the
 *      identical (allowed, remaining, retry_after_ms) sequence Go's
 *      ShardedLimiter produces — proving the fullAt encoding is behaviorally
 *      equivalent to the {tokens, last} encoding, not just "close enough."
 *   2. An architectural on-ramp, not a shipped optimization. Node CAN run
 *      genuinely parallel code via `worker_threads`, and a `SharedArrayBuffer`
 *      + `Atomics` backed bucket array WOULD make cross-worker rate limiting
 *      lock-free for real. This class does NOT attempt that — it would be
 *      substantial additional work (a shared memory layout, worker-safe
 *      shard hashing, Atomics.compareExchange loops) that nothing here
 *      requires today. What this class does do is keep each bucket's entire
 *      state as one encoded number (`fullAt`), the same representation Go
 *      uses, specifically so that door stays open: swapping the `Map<string,
 *      { fullAt: number }>` backing store for a `Float64Array` view over a
 *      `SharedArrayBuffer` is a storage-layer change, not an algorithm
 *      rewrite, if someone builds that later.
 *
 * Algorithm (ms clock, matching Node's Clock interface; Go uses ns):
 *
 *   tokens(now)   = burst − max(0, fullAt − now) / 1000 × rps
 *   consume n     = newFullAt = now + (burst − (tokens − n)) / rps × 1000
 *   zero-value fullAt (0) encodes a full bucket — a freshly seen key needs
 *   no initialization write, and (unlike RateLimiter's explicit 10-minute
 *   idle reset) an arbitrarily idle bucket refills to full by construction:
 *   the math above saturates at `burst` once the deficit clears.
 *   deny: retryAfterMs = max(1000, ceil((n − tokens) / rps) × 1000) — the
 *   same whole-second-ceil rounding RateLimiter's incrementLocal/peek use
 *   (see rate-limiter.ts), asserted for both SDKs by the conformance vectors.
 *
 * Storage note: unlike Go's `boundedCache`-backed shards, this port uses a
 * plain, unbounded `Map` per shard — the point of this class is decision
 * parity for conformance testing, not production memory bounding. Callers
 * who need bounded key-space memory should use the default `RateLimiter`
 * (which already wraps a `BoundedCache`).
 *
 * Source: token bucket — https://en.wikipedia.org/wiki/Token_bucket
 */

import type { BucketState, Clock, RateLimitDecision, RateLimitOptions } from '../types.js';
import type { RateLimiterLike } from './rate-limiter.js';

/** Shard count — must be a power of two. Matches Go's `shardCount`. */
export const SHARDED_LIMITER_SHARD_COUNT = 64;

interface AtomicBucket {
  /** The instant (ms, per this.clock) at which the bucket would be full. */
  fullAt: number;
}

class LimiterShard {
  readonly buckets = new Map<string, AtomicBucket>();
}

/**
 * Lock-free-in-Go, plain-in-Node sharded token bucket limiter. See the
 * file-level doc comment for why this exists despite Node's single-threaded
 * execution model removing the contention problem Go's version solves.
 */
export class ShardedLimiter implements RateLimiterLike {
  private readonly clock: Clock;
  private readonly shards: LimiterShard[];

  constructor(options: { clock: Clock }) {
    this.clock = options.clock;
    this.shards = Array.from({ length: SHARDED_LIMITER_SHARD_COUNT }, () => new LimiterShard());
  }

  /**
   * Hashes key with an inlined 32-bit FNV-1a (zero allocations) and masks
   * into the shard array. Matching Go's 64-bit FNV-1a exactly is NOT
   * required — the shard count and hash function are internal implementation
   * details. What must match Go is the observable (allowed, remaining)
   * decision sequence, not which shard a key happens to land in.
   * Source: FNV hash — http://www.isthe.com/chongo/tech/comp/fnv/
   */
  private shardFor(key: string): LimiterShard {
    let hash = 0x811c9dc5; // FNV offset basis (32-bit)
    for (let i = 0; i < key.length; i++) {
      hash ^= key.charCodeAt(i);
      hash = Math.imul(hash, 0x01000193); // FNV prime (32-bit)
    }
    const index = (hash >>> 0) & (SHARDED_LIMITER_SHARD_COUNT - 1);
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    return this.shards[index]!;
  }

  private bucketFor(key: string, create: boolean): AtomicBucket | undefined {
    const shard = this.shardFor(key);
    let bucket = shard.buckets.get(key);
    if (!bucket && create) {
      bucket = { fullAt: 0 }; // zero value = full bucket, no init write needed
      shard.buckets.set(key, bucket);
    }
    return bucket;
  }

  private tokensAt(fullAt: number, now: number, burst: number, rps: number): number {
    const deficitMs = fullAt - now;
    if (deficitMs <= 0) {
      return burst;
    }
    const tokens = burst - (deficitMs / 1000) * rps;
    return tokens < 0 ? 0 : tokens;
  }

  /** Mirrors RateLimiter's retry-after rounding exactly: whole seconds, floored at 1000ms. */
  private denyDecision(tokens: number, rps: number, need: number): RateLimitDecision {
    const retrySeconds = Math.ceil((need - tokens) / rps);
    const retryAfterMs = Math.max(1000, retrySeconds * 1000);
    return { allowed: false, applied: true, remaining: 0, retryAfterMs, limit: rps, degraded: false };
  }

  /** Allow is increment(key, options, 1) — no remote control-plane fallback exists at this layer. */
  allow(key: string, options: Required<RateLimitOptions> & { apiKey: string | undefined }): RateLimitDecision {
    return this.increment(key, options, 1);
  }

  /**
   * Consumes n tokens. Used when a single call costs more than one unit of
   * the limit (e.g. an LLM request billed by estimated token count).
   */
  increment(key: string, options: Required<RateLimitOptions>, n: number): RateLimitDecision {
    const rps = options.requestsPerSecond;
    const burst = options.burst;
    if (rps <= 0 || burst <= 0) {
      return { allowed: true, applied: false, remaining: -1, retryAfterMs: 0, limit: -1, degraded: false };
    }

    const now = this.clock.now();
    const bucket = this.bucketFor(key, true);
    // bucketFor(key, true) always returns a value when create=true.
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    const b = bucket!;

    const tokens = this.tokensAt(b.fullAt, now, burst, rps);
    if (tokens < n) {
      // Denials do not consume and need no state write: the deficit already
      // encoded in fullAt represents the refilled state.
      return this.denyDecision(tokens, rps, n);
    }

    const newTokens = tokens - n;
    b.fullAt = now + ((burst - newTokens) / rps) * 1000;

    const remaining = Math.max(0, Math.floor(newTokens));
    return { allowed: true, applied: true, remaining, retryAfterMs: 0, limit: rps, degraded: false };
  }

  /** Reports what allow() would decide right now WITHOUT consuming a token. Never creates bucket state for unseen keys. */
  peek(key: string, options: Required<RateLimitOptions>): RateLimitDecision {
    const rps = options.requestsPerSecond;
    const burst = options.burst;
    if (rps <= 0 || burst <= 0) {
      return { allowed: true, applied: false, remaining: -1, retryAfterMs: 0, limit: -1, degraded: false };
    }

    const now = this.clock.now();
    const bucket = this.bucketFor(key, false);
    if (!bucket) {
      return { allowed: true, applied: true, remaining: burst, retryAfterMs: 0, limit: rps, degraded: false };
    }

    const tokens = this.tokensAt(bucket.fullAt, now, burst, rps);
    if (tokens < 1) {
      return this.denyDecision(tokens, rps, 1);
    }
    return { allowed: true, applied: true, remaining: Math.floor(tokens), retryAfterMs: 0, limit: rps, degraded: false };
  }

  /** Returns current bucket state without consuming anything. Never creates bucket state for unseen keys. */
  get(key: string, options: Required<RateLimitOptions>): BucketState {
    const rps = options.requestsPerSecond;
    const burst = options.burst;
    const bucket = this.bucketFor(key, false);
    if (!bucket) {
      return { tokens: burst, capacity: burst, limit: rps };
    }

    const now = this.clock.now();
    const tokens = this.tokensAt(bucket.fullAt, now, burst, rps);
    return { tokens, capacity: burst, limit: rps };
  }

  /** Clears key's bucket; the next access starts from a full bucket. */
  reset(key: string): void {
    this.shardFor(key).buckets.delete(key);
  }
}
